npep/npep-122.md

```diff
@@ -0,0 +1,194 @@
+    Higher priority to allow/deny some inter-tenant connections, for example, allow ingress traffic from the
+    monitoring namespace without including that namespace to every tenant. Lower priority to allow/deny
+    connections that are not a part of tenant policy, for example, allow internal tenant connections with tenancy
+    and deny-all policy with lower priority.
+
+#### Story 4.5: Ensure every namespace in the cluster is a part of some tenant
+
+    As a cluster admin, I want to implement security zones tenants, and deny connections between different zones.
+    I want to use a label e.g. `security-zone` to build tenants, and I want to make sure every namespace in the cluster
+    is a part of some tenant/security zone. This way, I can be sure that even if the `security-zone` 
+    label is not assigned to the tenant for some reason, tenancy rules will still apply and security zones won't be
+    violated.
+
+#### What I couldn't figure our user stories for
+
+- Skip action
+- Ports *[]AdminNetworkPolicyPort
+- Multiple tenant policies (e.g. based on different labels?)
+
+### Existing API
+
+AdminNetworkPolicy has the “SameLabels” and “NotSameLabels” fields to support the use cases involving tenancy. For example:
+
+**Use case**
+
+Traffic should be disallowed by default between namespaces owned by different users (defined by labels i.e `user=foo` or `user=bar`).
+
+**Policy**
+```
+kind: BaselineAdminNetworkPolicy
+apiVersion: policy.networking.k8s.io/v1alpha1
+metadata:
+  name: user-based-tenancy
+spec:
+  subject:
+    namespaces:
+      matchExpressions:
+        - key: user
+          operator: Exists
+  ingress:
+    - action: Deny
+      from:
+        - namespaces:
+            notSameLabels:
+              - user
+```
+
+**Meaning**
+
+In namespaces that have a “user” label, by default, deny ingress from namespaces that have a different “user” label.
+
+There are two major problems with this implementation of tenancy as it pertains to the user stories.
+
+First, There is no explicit definition of "tenancy" anywhere. The administrator has an idea of 
+"tenants are defined by the user label", but that's only true because this particular ANP happens to include that
+particular rule, and there's no way to find the ANP(s) that defines tenancy if you don't already know what they are.
+
+Second, the SameLabels/NotSameLabels selectors behave really differently from other peers, causing multiple underlying 
+rules to be created, and the syntax doesn't make that obvious.
+
+Third, the syntax is very general purpose / powerful. ANP has subjects and peers, which are different, 
+and currently Tenancy is defined on the peers side. Tenancy by itself has the same subject and peer, 
+at least for the existing use cases, and having separate selectors for subject and peer allows for more
+configurations than needed. 
+
+Fourth, the ANP subject allows using pod selectors, while tenancy use cases only need namespace selectors.
```

Comment:
Is this a hard requirement? If a admin wants to isolate Pods by a label is that a valid use-case? Maybe we can add it, but I don't think thats too unreasonable and I don't think it is unimplementable (citation needed).

For example: I want to define my tenant as the value of the label `app` and deny Pods with that label from trying to talk to anything outside their tenant.

Reply:
It should be implementable indeed, but that adds extra complexity for the API and we don't have any use cases for it yet. Another concern is that pod labels can usually be changed by the namespace owner, in that case admin can't enforce proper tenancy isolation base don pod labels.

Reply:
> If a admin wants to isolate Pods by a label is that a valid use-case?

But this isn't about merely isolating Pods, it's about defining _tenancy_ at the Pod level. It's one thing to have an ANP that says "the frob operator should be able to access any pod with the 'frobbable' label in any Namespace". But it would be weird to say "all pods with the 'pod-tenant' label can talk to all other pods with the same valued 'pod-tenant' label in all namespaces", particularly since, as Nadia said, the users can just relabel their pods themselves whenever they want.